watchtower/server/server.go

```diff
@@ -0,0 +1,550 @@
+	peer, ok := c.(Peer)
+	if !ok {
+		log.Warnf("incoming connection %T does not satisfy "+
+			"server.Peer interface", c)
+		c.Close()
+		return
+	}
+
+	s.InboundPeerConnected(peer)
+}
+
+// InboundPeerConnected accepts a server.Peer, and handles the request submitted
+// by the client. This method serves also as a public endpoint for locally
+// registering new clients with the server.
+func (s *server) InboundPeerConnected(peer Peer) {
+	s.wg.Add(1)
+	go s.handleClient(peer)
+}
+
+// handleClient processes a series watchtower messages sent by a client. The
+// client may either send:
+//  * a single SessionInit message.
+//  * a series of StateUpdate messages.
+//
+// This method uses the server's peer map to ensure at most one peer using the
+// same session id can enter the main event loop. The connection will be
+// dropped by the watchtower if no messages are sent or received by the
+// configured Read/WriteTimeouts.
+//
+// NOTE: This method MUST be run as a goroutine.
+func (s *server) handleClient(peer Peer) {
+	defer s.wg.Done()
+
+	// Use the connection's remote pubkey as the client's session id.
+	id := wtdb.NewSessionIDFromPubKey(peer.RemotePub())
+
+	// Register this peer in the server's client map, and defer the
+	// connection's cleanup. If the peer already exists, we will close the
+	// connection and exit immediately.
+	err := s.addPeer(&id, peer)
+	if err != nil {
+		peer.Close()
+		return
+	}
+	defer s.removePeer(&id)
+
+	log.Infof("Accepted incoming peer %s@%s",
+		id, peer.RemoteAddr())
+	defer log.Infof("Releasing incoming peer %s@%s",
+		id, peer.RemoteAddr())
+
+	// stateUpdateOnlyMode will become true if the client's first message is
+	// a StateUpdate. If instead, it is a SessionInit, this method will exit
+	// immediately after replying. We track this to ensure that the client
+	// can't send a SessionInit after having already sent a StateUpdate.
+	var stateUpdateOnlyMode bool
+	for {
+		select {
+		case <-s.quit:
+			return
+		default:
+		}
+
+		// Set a read timeout to ensure we drop the client if not sent
+		// in a timely manner.
+		err = peer.SetReadDeadline(time.Now().Add(s.cfg.ReadTimeout))
+		if err != nil {
+			log.Errorf("unable to set read deadline: %v", err)
+			return
+		}
+
+		// Pull the next message off the wire, and parse it according to
+		// the watchtower wire specification.
+		rawMsg, err := peer.ReadNextMessage()
+		if err != nil {
+			log.Errorf("unable to read message: %v", err)
+			return
+		}
+
+		msgReader := bytes.NewReader(rawMsg)
+		nextMsg, err := wtwire.ReadMessage(msgReader, 0)
+		if err != nil {
+			log.Errorf("unable to parse message: %v", err)
+			return
+		}
+
+		// Process the request according to the message's type.
+		switch msg := nextMsg.(type) {
+
+		// A SessionInit indicates a request to establish a new session
+		// with our watchtower.
+		case *wtwire.SessionInit:
+			// Ensure SessionInit can only be sent as the first
+			// message.
+			if stateUpdateOnlyMode {
```

Comment:
This condition enforces that a SessionInit is only sent as the first message, and cannot follow a StateUpdate. For a single conn, we permit:
 - 1 `SessionInit`
 - n `StateUpdates`

When parsing the first message, we don't know (without querying the db) which case we're dealing with. After finishing the first loop, this becomes true to prevent an ordering of `StateUpdate`+, `SessionInit`.

If the connection expires, the client can resume an existing session using the same session keypair. Rationale for not having long-lived connections is it makes it more difficult for the tower to track when clients are on/offline, and also prevents clients from consuming tower resources endlessly.